File: LISTAS_URI/lista7/back_to_the_future.py
from collections import defaultdict
import sys

def augment(v, min_edge):
  if predecessor[v-1] != None:
    for u in graph[predecessor[v-1]]:
        min_edge = min(min_edge, graph[predecessor[v-1]][u]["flow"])

  if predecessor[v-1] != None:
    for u in graph[predecessor[v-1]]:
      graph[predecessor[v-1]][u]["flow"] -= min_edge
      graph[predecessor[v-1]][u]["flow"] += min_edge

  return min_edge

def dijkstra(graph, source, terminal):
  distances = [sys.maxsize] * (max(graph)+1)
  visited = [False] * (max(graph)+1)

  queue = []
  distances[source] = 0
  visited[source] = True
  queue.append(source)

  while queue:
    output = queue.pop(0)

    for vertex in graph[output]:
      if graph[output][vertex]["flow"] and distances[output] + graph[output][vertex]["cost"] < distances[vertex]:
        distances[vertex] = distances[output] + graph[output][vertex]["cost"]
        predecessor[vertex] = output

        if visited[vertex] == False:
          visited[vertex] = True
          queue.append(vertex)

    visited[output] = False
  
  sp = distances[terminal]
  result = sp != sys.maxsize
  return result, sp

edges = []
predecessor = []
instance = 1
d = 0
k = 0
f = 0
while True:
  n, m = [int(i) for i in input().split(" ")]
  graph = {}

  a = 0
  previous = [-1 for i in range(1, n+1)]

  for i in range(1, m+1):
    a, b, c = [int(i) for i in input().split(" ")]
    edges.append({"a": a, "b": b, "c": c})

  d, k = [int(i) for i in input().split(" ")]

  graph[0] = ({1: {"cost": 0, "flow": d}})
  graph[n] = ({(n+1): {"cost": 0, "flow": d}})

  for i in range(m):
    graph[edges[i]["a"]] = ({edges[i]["b"]: {"cost": edges[i]["c"], "flow": k}})
    graph[edges[i]["b"]] = ({edges[i]["a"]: {"cost": edges[i]["c"], "flow": k}})
  
  predecessor = [None] * (len(graph)+1)
  max_flow = 0
  resp = 0

  result, sp = dijkstra(graph, 0, n)
  while result:
    f = augment(n+1, sys.maxsize)
    max_flow += f
    resp += sp * f

    if max_flow == d:
      break

    result, sp = dijkstra(graph, 0, n)
  
  print("Instancia", instance)
  if max_flow != d:
    print("impossivel\n")
  else:
    print(resp, "\n")
  
  instance += 1

File: LISTAS_URI/lista7/backToTheFuture.cpp
#include <climits>
#include <cstdio>
#include <queue>
#include <vector>

#define MAX 1500
#define MAXN 21000
#define INFINITY INT_MAX

using namespace std;

struct edge {
    int a, b;
    int c;
};
struct graphEdge {
    int u, v, next;
    int cost, flow;
    graphEdge(int u = 0, int v = 0, int next = 0, int cost = 0, int flow = 0)
        : u(u)
        , v(v)
        , next(next)
        , cost(cost)
        , flow(flow)
    {
    }
};

int V, E, A, previous[MAXN];
int flow, sp, dist[MAX], p[MAX];
bool visited[MAX];
vector<graphEdge> graph(MAXN);
vector<edge> edges(MAXN);

void addEdge(int u, int v, int c, int f) {
    graph[A] = graphEdge(u, v, previous[u], c, f);
    previous[u] = A++;
    graph[A] = graphEdge(v, u, previous[v], -c, 0);
    previous[v] = A++;
}

void findFlow(int v, int minEdge) {
    for (int i = p[v]; i != -1; i = p[graph[i].u])
        minEdge = min(minEdge, graph[i].flow);

    for (int i = p[v]; i != -1; i = p[graph[i].u]) {
        graph[i].flow -= minEdge;
        graph[i ^ 1].flow += minEdge;
    }
    flow = minEdge;
}

bool dijkstra(int s, int t) {
    for (int i = 0; i <= V + 1; ++i) {
        dist[i] = INFINITY;
        p[i] = -1;
        visited[i] = false;
    }
    queue<int> q;
    dist[s] = 0;
    visited[s] = true;
    q.push(s);

    while (!q.empty()) {
        int u = q.front();
        q.pop();
        for (int i = previous[u]; i != -1; i = graph[i].next) {
            int v = graph[i].v;
            if (graph[i].flow && dist[u] + graph[i].cost < dist[v]) {
                dist[v] = dist[u] + graph[i].cost;
                p[v] = i;
                if (!visited[v]) {
                    visited[v] = true;
                    q.push(v);
                }
            }
        }
        visited[u] = false;
    }
    sp = dist[t];
    return sp != INFINITY;
}

int main() {
    int instance = 1;
    int d, k;
    while (scanf("%d %d", &V, &E) != EOF) {
      A = 0;
      for (int i = 0; i <= V + 1; ++i)
        previous[i] = -1;
      for (int i = 0; i < E; ++i)
        scanf("%d %d %d", &edges[i].a, &edges[i].b, &edges[i].c);
      scanf("%d %d", &d, &k);
      addEdge(0, 1, 0, d);
      addEdge(V, V + 1, 0, d);
      for (int i = 0; i < E; ++i) {
        addEdge(edges[i].a, edges[i].b, edges[i].c, k);
        addEdge(edges[i].b, edges[i].a, edges[i].c, k);
      }
      int maxFlow = 0, resp = 0;
      while (dijkstra(0, V + 1)) {
        flow= 0;
        findFlow(V + 1, INFINITY);
        maxFlow += flow;
        resp += sp * flow;

        if (maxFlow == d) break;
      }
      
      printf("Instancia %d\n", instance++);
      maxFlow != d ? printf("impossivel\n\n") : printf("%d\n\n", resp);
    }
  return 0;
}
